### PR TITLE
Refactored command handling code.

### PR DESCRIPTION
**Pull request by rbryan**
Refactored command handling code. Commands are now handled in cmd.c.

### CODE REVIEW

File: cmd.c [hunk 1]
```diff
@@ -0,0 +1,1842 @@
...
+CMD_FUN (mode);
+CMD_FUN (cmd);
+CMD_FUN (help);
+
+//Number of Commands
+#define NCMDS 33
+
+//cmd_funcs can't be malloc'd the regular way because of -Werror
+//malloc returns a void pointer and with -Werror you can't
+//cast from a regular pointer to a function pointer. -Russell
```
Comment (ec429) [inline cmd.c]: Could you not use `int (**cmd_funcs)(args...) = malloc(n * sizeof(int (*)(args...)))`?  A function pointer is data, so a pointer to function pointer should be a data pointer and thus castable from `void *`.  (Just tested a tiny program and it compiles happily.)
Agreed you can't malloc an array of functions, but an array of function pointers is fine.

"Any problem in Computer Science can be solved by another layer of indirection."

File: cmd.c [hunk 2]
```diff
@@ -0,0 +1,1842 @@
...
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-but-set-parameter"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+CMD_FUN (help)
+{
+	if(args)
+	{
+		int i = get_cmd_index(args);
+		fprintf(stderr, "%d\n",i);
```
Comment (ec429) [inline cmd.c]: Is this left-behind debugging code?

**Comment by ec429** [comment]
I like the general idea, but rather than separate arrays for commands, cmd_funcs and help, I think you should have an array of `struct {char *command; int (*cmd_func)(blah...); char *help;}`.

**Comment by rbryan** [comment]
I'm pretty sure I tried the bit with function pointers and it didn't work for the same or similar reasons. In any case it's rendered a non-issue by putting everything into a structure which definitely uses a regular pointer.
The pragmas turning of warnings aren't debugging code. Because I'm using an array of function pointers (or structures containing them) all of the functions have to take exactly the same set of arguments. Because some of the functions don't use all of the arguments they are given there are more problems with warnings and -Werror. I can't think of any way to use this sort of structure and get around this. You could maybe make some sort of function that does "something" without really doing anything so that the compiler is happy but that just doesn't feel right and it wouldn't exactly help the readability of the code.
I think you're right about using structures. No clue why I didn't just do that first. I'll implement it and push the changes.

**Comment by ec429** [comment]
By "debugging code" I was referring to the fprintf to stderr (line 183).  For some reason, GitHub thinks I'm commenting on a block of lines ending on the line I'm actually commenting on.
I agree about the unused params; normally I'd use `__attribute__((unused))` but the declarations are coming from a macro which precludes that.  There's always `(void)foo;` but that's an ugly hack…
Looking forward to the updated pull request.

**Comment by rbryan** [comment]
Yeah, that was debugging code. It should be gone now. I've pushed the new version that uses an array of structs instead of separate arrays.

**Comment by rbryan** [comment]
Out of curiosity, Is there any way to highlight the users nick when it's mentioned? Also, what's the status on scripting?

**Comment by ec429** [comment]
I decided the remaining changes I wanted to see were easier to do than describe.  So, I've pulled into a branch [rbryan-commands](https://github.com/ec429/quIRC/tree/rbryan-commands) and made my changes on top of it.  Take a look, see if anything seems objectionable.  If not I'll merge into master, and then probably cut a release.

**Comment by ec429** [comment]
As for your recent questions:
* Yes, it should be possible to highlight the nick.  My favoured approach would be to make render_line() pass a bool to wordline(), true for 'message' and false for 'stamp' and 'tag', which tells wordline() to highlight any word that matches the nick.  I may even attempt to do this in the near future (but, wordline() is scary).
* Scripting has been stalled for a long time because I've been too busy with other projects to work on quirc.  But now you've got me interested in quirc again, so it might happen...

**Comment by rbryan** [comment]
Everything looks great to me. I don't see anything that I don't think is an obvious improvement.

I'll look into nick highlighting if I can find time. That's a feature that I would find very useful.

Were you planning on embedding a scripting language or rolling your own?

One other quick thought. This may make more sense if and when a scripting system is added, but it might be a good idea to have some sort of command hierarchy. Say for instance someone tries the command "/foo". quIRC fails to find foo in the command array so it looks for a script to call. Having found no script quIRC might try to send it as a server command. Right now this probably doesn't make too much sense because any unidentified command would either result in an error response from the server or a mess of text if they tried something like "/list".

**Comment by ec429** [comment]
The plan for scripting is to use an application protocol; basically you have a pipe with quIRC at one end and a script at the other.  The protocol is simple and textual so scripts can be written in more or less anything - the scripts I've used for testing are in `awk`!
The work on this is in the [symbiont](https://github.com/ec429/quIRC/tree/symbiont) branch.  Scripts can declare themselves interested in received traffic on a buffer, and can "grab" a command, either creating a new command or intercepting an existing one.

**Pull request merged**